### ISSUE TITLE
Code Quality: Refactor Plugin class

### ISSUE THREAD

**Issue by spacedmonkey** [GoogleForCreators/web-stories-wp#5292]
There currently is a PHP in Plugin class has an ignore comment for the PHP MD as the class has too many properties. We should refactor the plugin class, as it doesn't use any many class properties.

**Comment by spacedmonkey** [comment]
When thinking over this ticket, I thought we could do something like this. 

```
		// Embed base.
		$embed_base = new Embed_Base();
		add_action( 'init', [ $embed_base, 'init' ], 9 );
                $this->register['embed_base'] = $embed_base;

		// Gutenberg Blocks.
		$embed_block = new Embed_Block();
		add_action( 'init', [ $embed_block, 'init' ] );
                $this->register['block']['embed_block] = $embed_block;

		// Embed shortcode.
		$embed_shortcode = new Embed_Shortcode();
		add_action( 'init', [ $embed_shortcode, 'init' ] );
                $this->register['shortcode']['embed_shortcode] = $embed_shortcode;
```

All classes on the root of the namespace, use the class name as the key. All classes with a different namespace, like `\Google\Web_Stories\Blocks` would be under the `['block']` key. 

The for backwards compatibility, use a magic getter. 

```
	public function __get( $key ) {
		switch ( $key ) {
			case 'database_upgrader':
				return $this->register['database_upgrader'];
			case 'embed_shortcode':
				return $this->register['shortcode']['embed_shortcode];
			case 'story':
				return $this->register['story'];
			default: 
				if ( isset( $this->$key ) ) {
					return $this->$key;
				}
		}

		return null;
	}
```

**Comment by swissspidy** [comment]
> All classes on the root of the namespace, use the class name as the key. 

That only scales to a certain point. With more namespaces likely being added in the future (blocks, shortcodes, customizer integrations, stuff like that), we'll quickly run into the same problems again regarding code bloat.

Maybe we can take inspiration from the AMP plugin here:

https://github.com/ampproject/amp-wp/blob/develop/src/AmpWpPlugin.php
https://github.com/ampproject/amp-wp/blob/develop/src/Infrastructure/ServiceBasedPlugin.php

**Comment by spacedmonkey** [comment]
I think that we could look at [Service Container](https://laravel.com/docs/8.x/container).

**Comment by spacedmonkey** [comment]
We could also look at some more simple like [pimple](https://pimple.symfony.com/).

**Comment by spacedmonkey** [comment]
We could also do something similar to amp plugin. 

```
const services = [
    'admin'  => ['class' => \Google\Web_Stories\Admin::class,  'action' => 'admin_init', 'method' => 'init', 'priority' => 10 ], 
    'media'  => ['class' => \Google\Web_Stories\Media::class,  'action' => 'init', 'method' => 'init', 'priority' => 10 ], 
    'tracking'  => ['class' => \Google\Web_Stories\Tracking::class,  'action' => 'init', 'method' => 'init', 'priority' => 10 ], 
    'template'  => ['class' => \Google\Web_Stories\Template_Post_Type::class,  'action' => 'init', 'method' => 'init', 'priority' => 10 ], 	
      'story'  => ['class' => \Google\Web_Stories\Story_Post_Type::class,  'action' => 'init', 'method' => 'init', 'priority' => 10 ], 
]
```

Then register all the classes like this. 

```
public function register(){
    foreach( self:services as $name => $props ) {
         $this->object[$name] = new $props['class']();
         add_action( $props['action'], [ $this->object[$name], $props['method'], $props['priority'] );
    }
    
}
```

We would need to manage dependencies here. 

But we could do that with getters and setters like this. 

```
$experiments = new Experiments();
$this->object['story']->setExperiments($experiments);
```

**Comment by spacedmonkey** [comment]
As a talking point, I created a WIP PR with these idea https://github.com/google/web-stories-wp/pull/5467

**Comment by spacedmonkey** [comment]
Another cool example of service containers in PHP 

https://github.com/GraphQLAPI/graphql-api-for-wp/blob/master/src/Plugin.php

**Comment by spacedmonkey** [comment]
Closed via #6906
